add openfoam community packages for adiosWrite

olesenm opened the pull request.

davydden commented:
you may want to look at how extension packages are done for `matlab`, `octave`, `python`, etc. I think that's the correct way to do it with Spack.

I would also call it `openfoam-<extension>`  as opposed to `of-extension`, which is strange to read and you don't get those packages when searching for `openfoam`.

davydden commented on var/spack/repos/builtin/packages/of-iso-advector/package.py:
```diff
@@ -0,0 +1,153 @@
+                join_path(self.build_userdir, d),
+                join_path(self.prefix, d))
+
+# -----------------------------------------------------------------------------
```
i think we figured out that this line is not needed :smile:

olesenm replied:
Sure, but I don't have the feeling that this will make the merge happen any quicker.

olesenm commented:
I don't see how `extends()` works here - where are the files supposed to land? How will OpenFOAM ever know about them?

For naming we could have 'openfoam-' as a prefix, which means that openfoam-com and openfoam-org look like extensions rather than implementations?, or 'foam-', which means that foam-extend has the same problem..

davydden commented:
> I don't see how extends() works here - where are the files supposed to land?

such things land into `setup_dependent_package` of, for example, `octave` (the main package which is to be extended). 

Maybe that is not the best way to do it here, @adamjstewart @tgamblin what do you think?

As for the naming, there is a bit of confusion with `openfoam-com` and `openfoam-org`, but IMO it is still way better than `of-` which reads as preposition.

davydden commented:
@olesenm do those packages always extend `openfoam-com`? Can they be linked against `openfoam-org`?

If it's the former, maybe the packages should be called `openfoam-com-<package>`.

davydden commented:
@olesenm more questions to you: those files you add, i.e. `change-userdir.sh` and `spack-Allwmake`, are they almost the same for each package/extension of openfoam? `change-userdir.sh` looks identical, `spack-Allwmake` can probably be unified somehow. My point is, if we have another 100 extensions and each one adds its own `change-userdir.sh` and `spack-Allwmake` -- something is wrong.

olesenm commented:
Ideally the change-userdir.sh would land in central/common location such as openfoam-com/tools, where any other (openfoam) package could use it. Ultimately still have too many PR up in the air (eg, this PR and swak4foam will ultimately depend on the openfoam PR, but we wish to have the merges bring in smaller amounts each time).

The `spack-Allwmake` are fairly similar for the dependent packages, could probably also provide via the openfoam-com/tools too (when that gets merged)

olesenm commented:
Note that the packages _should_ general work with openfoam-com, openfoam-org and foam-extend, but  there will be some cases where they (temporarily?) only work with 1 or 2 providers/variants.

Maybe a `foam-` prefix would indeed make sense and provide reasonable grouping. Would 'foam-extend` be renamed as `openfoam-extend` for better grouping with the providers? But would this break existing spack installations? Or would it work to rejig openfoam-extend to also provides("foam-extend") ?
If so, this should go into that PR.

Possibilities
* of-swak4foam -> foam-swak  (or redundant foam-swak4foam?)
* of-adios-write -> foam-adios-write
* of-iso-advector -> foam-iso-advector
* of-extend -> openfoam-extend, provides 'foam-extend'

adamjstewart commented:
@davydden I think you're misunderstanding the purpose of extensions in Spack. The difference between `extends()` and `depends_on()` isn't just semantic, it's functional as well. If a package `extends()` another package, that allows you to run `spack activate` on that package. This is needed for things like Python, where it is convenient to symlink the module into the base Python installation so that you don't have to worry about environment variables. Basically, unless adios-write and iso-advector need to be symlinked into the OpenFOAM installation, it doesn't make sense to call them extensions. Likewise, there's no reason to rename the package if it isn't a real extension. `adios-write` and `iso-advector` are perfectly fine names.

davydden commented:
@adamjstewart thanks for clarification! I agree.

olesenm commented:
I will update this again in the next couple of weeks, so we can hold off again for a while.

olesenm commented:
@adamjstewart : now adjusted and updated. Now only has adios-write, the isoadvector has been integrated into OpenFOAM-1706 itself, so no need for an addon. I left the branch name the same for the merge, but it is now really just the adios bit.

adamjstewart commented on var/spack/repos/builtin/packages/of-adios-write/package.py:
```diff
@@ -0,0 +1,136 @@
+    homepage = "https://develop.openfoam.com/Community/feature-adiosWrite/"
+    gitrepo  = "https://develop.openfoam.com/Community/feature-adiosWrite.git"
+
+    version('plus', branch='develop', git=gitrepo)
```
It would be better to call this version `develop` as I mentioned in your other PR.

adamjstewart commented on var/spack/repos/builtin/packages/of-adios-write/package.py:
```diff
@@ -0,0 +1,136 @@
+    def install(self, spec, prefix):
+        """Install under the prefix directory"""
+
+        for f in ['README.md']:
```
Why the loop?

olesenm replied:
Thanks. There is/was an Issues file.

olesenm commented:
Hi @adamjstewart ,

I fully agree that `of-adios-write` looks really long-winded. Although it does not extend OpenFOAM in the python sense of things, it does provide additional libraries that _only_ work with OpenFOAM.
I figured that the `of-` prefix was ugly enough not to collide with another project that is also using ADIOS for writing.

adamjstewart commented:
I'm still not confident that this is the right name for the package, but we can rename it later if we change our minds.
